x-pack/plugins/upgrade_assistant/public/application/components/es_deprecations/deprecations/ml_snapshots/use_snapshot_state.tsx

```diff
@@ -0,0 +1,151 @@
+  jobId,
+  snapshotId,
+  api,
+}: {
+  jobId: string;
+  snapshotId: string;
+  api: ApiService;
+}) => {
+  const [requestError, setRequestError] = useState<ResponseError | undefined>(undefined);
+  const [snapshotState, setSnapshotState] = useState<SnapshotStatus>({
+    status: 'idle',
+    jobId,
+    snapshotId,
+  });
+
+  const pollIntervalIdRef = useRef<ReturnType<typeof setTimeout> | null>(null);
+  const isMounted = useRef(false);
+
+  const clearPollInterval = useCallback(() => {
+    if (pollIntervalIdRef.current) {
+      clearTimeout(pollIntervalIdRef.current);
+      pollIntervalIdRef.current = null;
+    }
+  }, []);
+
+  const updateSnapshotStatus = useCallback(async () => {
+    clearPollInterval();
+
+    const { data, error: updateStatusError } = await api.getMlSnapshotUpgradeStatus({
+      jobId,
+      snapshotId,
+    });
+
+    if (updateStatusError) {
+      setSnapshotState({
+        snapshotId,
+        jobId,
+        action: 'upgrade',
+        status: 'error',
+      });
+      setRequestError(updateStatusError);
+      return;
+    }
+
+    setSnapshotState(data);
+
+    // Only keep polling if it exists and is in progress.
+    if (data?.status === 'in_progress') {
+      pollIntervalIdRef.current = setTimeout(updateSnapshotStatus, POLL_INTERVAL);
+    }
+  }, [api, clearPollInterval, jobId, snapshotId]);
+
+  const upgradeSnapshot = useCallback(async () => {
+    setSnapshotState({
+      snapshotId,
+      jobId,
+      action: 'upgrade',
+      status: 'in_progress',
+    });
+
+    const { data, error: upgradeError } = await api.upgradeMlSnapshot({ jobId, snapshotId });
+
+    if (upgradeError) {
+      setRequestError(upgradeError);
+      setSnapshotState({
+        snapshotId,
+        jobId,
+        action: 'upgrade',
+        status: 'error',
+      });
+      return;
+    }
+
+    setSnapshotState(data);
+    updateSnapshotStatus();
+  }, [api, jobId, snapshotId, updateSnapshotStatus]);
+
+  const deleteSnapshot = useCallback(async () => {
+    setSnapshotState({
+      snapshotId,
+      jobId,
+      action: 'delete',
+      status: 'in_progress',
+    });
+
+    const { error: deleteError } = await api.deleteMlSnapshot({
+      snapshotId,
+      jobId,
+    });
+
+    if (deleteError) {
+      setRequestError(deleteError);
+      setSnapshotState({
+        snapshotId,
+        jobId,
+        action: 'delete',
+        status: 'error',
+      });
+      return;
+    }
+
+    setSnapshotState({
+      snapshotId,
+      jobId,
+      action: 'delete',
+      status: 'complete',
+    });
+  }, [api, jobId, snapshotId]);
+
+  useEffect(() => {
+    isMounted.current = true;
+
+    return () => {
+      isMounted.current = false;
+
+      // Clean up on unmount.
+      clearPollInterval();
+    };
+  }, [clearPollInterval]);
+
+  return {
+    snapshotState: {
+      ...snapshotState,
+      error: requestError,
+    },
+    upgradeSnapshot,
+    updateSnapshotStatus,
+    deleteSnapshot,
+  };
```

Comment:
Amazing job creating this hook, @alisonelizabeth! 👍 
I wonder though if that needs to be unit tested or it's enough that there are CITs covering UIs using this hook.

Reply:
That's a great point. I'd like to merge this in for now as I think we do have basic coverage from the CITs. I'll be evaluating test coverage during 7.15 overall across UA and will look into adding additional tests here in a separate PR.